README.pod
=pod

=head1 NAME

MooseX::ShortCut::BuildInstance - A shortcut to build Moose instances

=begin html

<a href="https://www.perl.org">
	<img src="https://img.shields.io/badge/perl-5.10+-brightgreen.svg" alt="perl version">
</a>

<a href="https://travis-ci.org/jandrew/MooseX-ShortCut-BuildInstance">
	<img alt="Build Status" src="https://travis-ci.org/jandrew/MooseX-ShortCut-BuildInstance.png?branch=master" alt='Travis Build'/>
</a>

<a href='https://coveralls.io/github/jandrew/MooseX-ShortCut-BuildInstance?branch=master'>
	<img src='https://coveralls.io/repos/github/jandrew/MooseX-ShortCut-BuildInstance/badge.svg?branch=master' alt='Coverage Status' />
</a>

<a href='https://github.com/jandrew/MooseX-ShortCut-BuildInstance'>
	<img src="https://img.shields.io/github/tag/jandrew/MooseX-ShortCut-BuildInstance.svg?label=github version" alt="github version"/>
</a>

<a href="https://metacpan.org/pod/MooseX::ShortCut::BuildInstance">
	<img src="https://badge.fury.io/pl/MooseX-ShortCut-BuildInstance.svg?label=cpan version" alt="CPAN version" height="20">
</a>

<a href='http://cpants.cpanauthors.org/dist/MooseX-ShortCut-BuildInstance'>
	<img src='http://cpants.cpanauthors.org/dist/MooseX-ShortCut-BuildInstance.png' alt='kwalitee' height="20"/>
</a>

=end html

=head1 SYNOPSIS

	#!/user/bin/env perl
	package Mineral;
	use Moose;
	use Types::Standard qw( Enum );

	has 'type' =>(
			isa => Enum[qw( Quartz Diamond Basalt Granite )],
			is => 'ro'
		);

	package Identity;
	use Moose::Role;

	has 'name' =>( is => 'ro' );

	use lib '../../../lib';
	use MooseX::ShortCut::BuildInstance qw( should_re_use_classes build_instance );
	should_re_use_classes( 1 );# To reuse build_instance
	use Test::More;
	use Test::Moose;

	# First build of instance
	my 	$paco = build_instance(
			package => 'Pet::Rock',
			superclasses =>['Mineral'],
			roles =>['Identity'],
			type => 'Quartz',
			name => 'Paco',
		);

	does_ok( $paco, 'Identity', 'Check that the ' . $paco->meta->name . ' has an -Identity-' );
	print'My ' . $paco->meta->name . ' made from -' . $paco->type . '- (a ' .
	( join ', ', $paco->meta->superclasses ) . ') is called -' . $paco->name . "-\n";

	# Next instance (If you don't want to call build_instance again)
	my $Fransisco = Pet::Rock->new(
		type => 'Diamond',
		name => 'Fransisco',
	);
	does_ok( $Fransisco, 'Identity', 'Check that the ' . $Fransisco->meta->name . ' has an -Identity-' );
	print'My ' . $Fransisco->meta->name . ' made from -' . $Fransisco->type . '- (a ' .
	( join ', ', $Fransisco->meta->superclasses ) . ') is called -' . $Fransisco->name . "-\n";

	# Another instance (reusing build_instance)
	my $Gonzalo = build_instance(
			package => 'Pet::Rock',
			superclasses =>['Mineral'],
			roles =>['Identity'],
			type => 'Granite',
			name => 'Gonzalo',
		);
	does_ok( $Gonzalo, 'Identity', 'Check that the ' . $Gonzalo->meta->name . ' has an -Identity-' );
	print'My ' . $Gonzalo->meta->name . ' made from -' . $Gonzalo->type . '- (a ' .
	( join ', ', $Gonzalo->meta->superclasses ) . ') is called -' . $Gonzalo->name . "-\n";
	done_testing();

    ##############################################################################
    # Output of SYNOPSIS
    # 01:ok 1 - Check that the Pet::Rock has an -Identity-
    # 02:My Pet::Rock made from -Quartz- (a Mineral) is called -Paco-
    # 01:ok 1 - Check that the Pet::Rock has an -Identity-
    # 02:My Pet::Rock made from -Diamond- (a Mineral) is called -Fransisco-
    # 01:ok 1 - Check that the Pet::Rock has an -Identity-
    # 02:My Pet::Rock made from -Granite- (a Mineral) is called -Gonzalo-
    # 03:1..3
    ##############################################################################

=head1 DESCRIPTION

This module is a shortcut to custom build L<Moose> class instances on the fly.
The goal is to compose unique instances of Moose classes on the fly using a single
function call with information describing required attributes, methods, inherited
classes, and roles as well as any desired instance settings.  This package will
check for and fill in any missing pieces as needed so that your call can either be
complex or very simple.  The goal is to provide configurable instance building
without stringing together a series of Class-E<gt>method( %args ) calls.

The package can also be used as a class factory with the L<should_re_use_classes
|/$MooseX::ShortCut::BuildInstance::re_use_classes> method.

Even though this is a Moose based class it provides a functional interface.

=head1 WARNING(S)

Moose (and I think perl 5) can't have two classes with the same name but
different guts coexisting! This means that if you build a class (package) name
on the fly while building an instance and then recompose a new class (package) with
the same name but different functionality (different attributes, methods, inherited
classes or roles) while composing a new instance on the fly then all calls
to the old instance will use the new class functionality for execution. (Usually
causing hard to troubleshoot failures).  This is also true if you re-build a
prebuilt class name inititally installed with 'use'.

MooseX::ShortCut::BuildInstance will warn if you overwrite named classes (packages)
built on top of another class (package) also built by MooseX::ShortCut::BuildInstance.
If you are using the 'build_instance' method to generate multiple instances of
the same class (by 'package' name) with different attribute settings but built
with the same functionality then you need to understand the purpose of the
L<$re_use_classes|/$MooseX::ShortCut::BuildInstance::re_use_classes> global variable.
An alternative to multiple calls straight to 'build_instance' is to call
L<build_class|/build_class( %args|\%args )> separately and then just call -E<gt>new
against the resulting class name over and over again.  Another alternative is to
leave the 'package' argument out of 'build_instance' and let this class create a
unique by-instance anonymous class/package name.

MooseX::ShortCut::BuildInstance will also warn (but not stop you) if you try to
overwrite a pre-loaded package initially installed with 'use' or 'require'.

The Types module in this package uses L<Type::Tiny> which can, in the
background, use L<Type::Tiny::XS>.  While in general this is a good thing you will
need to make sure that Type::Tiny::XS is version 0.010 or newer since the older
ones didn't support the 'Optional' method.

This package will clone the passed arguments to L<build_class|/build_class> and
L<build_instance|/build_instance> since the references are destructivly parsed.
If that is not what you want then use the method L<set_args_cloning
|/set_args_cloning( $bool )> to manage the desired process.  Where this is likley
to go south is if your passed arguments contain a deep perl data set or reference
that you want shared.  In this case clone only the bits you want cloned on the
script side.

=head1 Functions for Export

=head2 build_instance( %args|\%args )

=over

B<Definition:> This method is used to create a Moose instance on the fly.
I<It assumes that you do not have the class pre-built and will look for the
needed information to compose a new class as well.>  Basically this passes the
%args intact to L<build_class|/build_class( %args|\%args )> first.  All the
relevant class building pieces will be used and removed from the args and then
this method will run $returned_class_name->new( %remaining_args ) with what is
left.

B<Accepts:> a hash or hashref of arguments.  They must include the
necessary information to build a class.  I<(if you already have a class just
call $class-E<gt>new( %args ); instead of this method!)> This hashref can also
contain any attribute settings for the instance as well.  See
L<build_class|/build_class( %args|\%args )> for more information.

B<Returns:> This will return a blessed instance of your new class with
the passed attribute settings implemented.

=back

=head2 build_class( %args|\%args )

=over

B<Definition:> This function is used to compose a Moose class on the fly.  The
the goal is to allow for as much or as little class definition as you want to be
provided by one function call.  The goal is also to remove as much of the boilerplate
and logic sequences for class building as possible and let this package handle that.
The function begins by using the L<Moose::Meta::Class>-E<gt>class(%args) method.
For this part the function specifically uses the argument callouts 'package',
'superclasses', and 'roles'.  Any necessary missing pieces will be provided. I<Even
though L<Moose::Meta::Class>-E<gt>class(%args) allows for the package name to be called
as the first element of an odd numbered list this implementation does not.  To define
a 'package' name it must be set as the value of the 'package' key in the %args.>
This function then takes the following arguements; 'add_attributes', 'add_methods',
and 'add_roles_in_sequence' and implements them in that order.   The
implementation of these values is done with L<Moose::Util> 'apply_all_roles'
and the meta capability in L<Moose>.

B<Accepts:> a hash or hashref of arguments.  Six keys are stripped from the hash or
hash ref of arguments.  I<These keys are always used to build the class.  They are
never passed on to %remaining_args.>

=over

B<The first three key-E<gt>value pairs are consumed simultaneously>.  They are;

=over

B<package:> This is the name (a string) that the new instance of
a this class is blessed under.  If this key is not provided the package
will generate a generic name.  This will L<overwrite|/WARNING> any class
built earlier with the same name.

=over

B<accepts:> a string

=back

B<superclasses:> this is intentionally the same key from
Moose::Meta::Class-E<gt>create.

=over

B<accepts:> a recognizable (by Moose) class name

=back

B<roles:> this is intentionally the same key from Moose::Meta::Class
-E<gt>create.

=over

B<accepts:> a recognizable (by Moose) class name

=back

=back

B<The second three key-E<gt>value pairs are consumed in the following
sequence>.  They are;

=over

B<add_attributes:> this will add attributes to the class using the
L<Moose::Meta::Class>-E<gt>add_attribute method.  Because these definitions
are passed as key / value pairs in a hash ref they are not added in
any specific order.

=over

B<accepts:> a hash ref where the keys are attribute names and the values
are hash refs of the normal definitions used to define a Moose attribute.

=back


B<add_methods:>  this will add methods to the class using the
L<Moose::Meta::Class>-E<gt>add_method method.  Because these definitions
are passed as key / value pairs in a hash ref they are not added in
any specific order.

=over

B<accepts:> a hash ref where the keys are method names and the values
are anonymous subroutines or subroutine references.

=back

B<add_roles_in_sequence:> this will compose, in sequence, each role in
the array ref into the class built on the prior three arguments using
L<Moose::Util> apply_all_roles.  This will allow an added role to
'require' elements of a role earlier in the sequence.  The roles
implemented with the L<role|/roles:> key are installed first and in a
group. Then these roles are installed one at a time.

=over

B<accepts:> an array ref list of roles recognizable (by Moose) as roles

=back

=back

=back

B<Returns:> This will check the caller and see if it wants an array or a
scalar.  In array context it returns the new class name and a hash ref of the
unused hash key/value pairs.  These are presumably the arguments for the
instance.  If the requested return is a scalar it just returns the name of
the newly created class.

=back

=head2 should_re_use_classes( $bool )

=over

This sets/changes the global variable
L<MooseX::ShortCut::BuildInstance::re_use_classes
|/$MooseX::ShortCut::BuildInstance::re_use_classes>

=back

=head2 set_class_immutability( $bool )

=over

This sets/changes the global variable
L<MooseX::ShortCut::BuildInstance::make_classes_immutable
|/$MooseX::ShortCut::BuildInstance::make_classes_immutable>

=back

=head2 set_args_cloning( $bool )

=over

This sets/changes the global variable
L<MooseX::ShortCut::BuildInstance::should_clone_args
|/$MooseX::ShortCut::BuildInstance::should_clone_args>

=back

=head1 GLOBAL VARIABLES

=head2 $MooseX::ShortCut::BuildInstance::anonymous_class_count

This is an integer that increments and appends to the anonymous package name
for each new anonymous package (class) created.

=head2 $MooseX::ShortCut::BuildInstance::built_classes

This is a hashref that tracks the class names ('package's) built buy this class
to manage duplicate build behaviour.

=head2 $MooseX::ShortCut::BuildInstance::re_use_classes

This is a boolean (1|0) variable that tracks if the class should overwrite or
re-use a package name (and the defined class) from a prior 'build_class' call.
If the package name is overwritten it will L<cluck|https://metacpan.org/pod/Carp#SYNOPSIS>
in warning since any changes will affect active instances of prior class builds
with the same name.  If you wish to avoid changing old built behaviour at the risk
of not installing new behaviour then set this variable to true.  I<No warning will
be provided if new requested class behaviour is discarded.> The class reuse behaviour
can be changed with the exported method L<should_re_use_classes
|/should_re_use_classes( $bool )>.  This does not apply to pre-loaded classes.
For pre-loaded classes this package will cluck and then overwrite every time.

=over

B<Default:> False = warn then overwrite

=back

=head2 $MooseX::ShortCut::BuildInstance::make_classes_immutable

This is a boolean (1|0) variable that manages whether a class is immutabilized at the end of
creation.  This can be changed with the exported method L<set_class_immutability
|/set_class_immutability( $bool )>.

=over

B<Default:> True = always immutabilize classes after building

=back

=head2 $MooseX::ShortCut::BuildInstance::should_clone_args

This is a boolean (1|0) variable that manages whether a the arguments passed to
L<build_instance|/build_instance( %argsE<verbar>\%args )> and L<build_class
|/build_class( %argsE<verbar>\%args )> are cloned (using L<Clone> )  the arguments
to both of these are processed destructivly so generally you would want them cloned
but not in every case.  If you want cloning to be managed on the script side set this
global variable to 0.  Where this is likley to be helpful is if your passed arguments
contain a deep perl data set or reference that you want shared.  In this case clone only
the bits you want cloned on the script side.

=over

B<Default:> True = always clone arguments

=back

=head1 Build/Install from Source

=over

B<1.> Download a compressed file with the code

B<2.> Extract the code from the compressed file.

=over

If you are using tar this should work:

	tar -zxvf Spreadsheet-XLSX-Reader-LibXML-v0.xx.tar.gz

=back

B<3.> Change (cd) into the extracted directory

B<4.> Run the following

=over

(For Windows find what version of make was used to compile your perl)

	perl  -V:make

(for Windows below substitute the correct make function (s/make/dmake/g)?)

=back

	>perl Makefile.PL

	>make

	>make test

	>make install # As sudo/root

	>make clean

=back

=head1 SUPPORT

=over

L<MooseX-ShortCut-BuildInstance/issues|https://github.com/jandrew/MooseX-ShortCut-BuildInstance/issues>

=back

=head1 TODO

=over

B<1.> L<Increase test coverage
|https://coveralls.io/github/jandrew/MooseX-ShortCut-BuildInstance?branch=master>

B<2.> Add an explicit import call as an action key in 'build_class'

=back

=head1 AUTHOR

=over

Jed Lund

jandrew@cpan.org

=back

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

This software is copyrighted (c) 2012, 2016 by Jed Lund

=head1 Dependencies

=over

L<version>

L<5.010|http://perldoc.perl.org/perl5100delta.html> (for use of
L<defined or|http://perldoc.perl.org/perlop.html#Logical-Defined-Or> //)

L<Moose>

L<Moose::Meta::Class>

L<Carp> - cluck

L<Moose::Exporter>

L<Moose::Util> - apply_all_roles

L<Moose::Exporter>

L<Type::Tiny> - 1.000

L<Data::Dumper>

L<MooseX::ShortCut::BuildInstance::Types>

=back

=head1 SEE ALSO

=over

L<Moose::Meta::Class> ->create

L<Moose::Util> ->with_traits

L<MooseX::ClassCompositor>

L<Log::Shiras|https://github.com/jandrew/Log-Shiras>

=over

All lines in this package that use Log::Shiras are commented out

=back

=back

=cut


Changes
Revision history for {{$dist->name}}

Uses the change tracking convention
+ New feature.
- Removed/depricated feature.
! Bug fix.
* Place holder

{{$NEXT}}
	
	* Another cool change

v1.44.2   2016-08-16 19:11:16-07:00 America/Los_Angeles
	
	! Removed the direct call to Log::Shiras::Switchboard since I couldn't solve the recursive inheritance fails
		use Log::Shiras::TapWarn to feed internal stuff to logs

v1.42.0   2016-08-12 13:12:42-07:00 America/Los_Angeles

	+ Updated the module ::BuildInstance to have new targeted Log::Shiras debug lines
	+ Removed the UnhideDebug module in favor of the '~::Unhide' in Log::Shiras (soon to be released)

v1.40.2   2016-05-01 09:41:57-07:00 America/Los_Angeles
	
	+ Added the global variable $should_clone_args and a method to set it
	+ Added some additional debugging through Log::Shiras

v1.38.4   2016-04-12 22:41:13-07:00 America/Los_Angeles
	
	! re-added MooseX::ShortCut::BuildInstance::UnhideDebug I still use it for debugging
		The source filter is still not run till $ENV{log_shiras_filter_on} is set

v1.38.2   2016-04-12 17:14:35-07:00 America/Los_Angeles

	! Cleaned up a left-in debug print statement

v1.36.8   2016-04-12 16:36:52-07:00 America/Los_Angeles
	
	! Fixed implementation of the re_use_classes to include all classes

v1.36.6   2016-04-12 14:53:22-07:00 America/Los_Angeles
	
	! Fixed the badges
	+ Update the .travis.yml file

v1.36.4   2016-04-12 14:39:22-07:00 America/Los_Angeles

	! Really removed MooseX::ShortCut::BuildInstance::UnhideDebug
		after it got repopulated with a sync from Github

v1.36.2   2016-04-12 14:31:59-07:00 America/Los_Angeles

	- Removed the class MooseX::ShortCut::BuildInstance::UnhideDebug (Not actually used recently)
	+ Added the ability to overwrite pre-built classes loaded with use
		It does warn before doing so.

v1.34.8   2015-10-27 13:09:54-07:00 America/Los_Angeles

	+ A minor change to the logging comments for build fails
	+ changed #! perl to #!/usr/bin/env perl in the SYNOPSIS

v1.34.6   2015-09-25 06:01:53-07:00 America/Los_Angeles

	! Added the configuration requirement for Moose 2.1213 before test since I got a weird fail for older moose when it was listed as installed
	+ Added clone to the args so that passed references would stay safe for re-use
	+ Additional badge changes

v1.34.4   2015-03-31 05:30:37-07:00 America/Los_Angeles

	+ Added a .travis.yml file for TravisCI coverage and Coveralls coverage
	! Improved the manifest skip file
	+ Added badges for the README
	+ Added a Type::Tiny 1.000 requirement from 0.46
	+ Reordered the Changes file to newest to oldest

v1.34.2   2014-12-24 20:51:33-08:00 America/Los_Angeles

	+ Improved error messaging for error instance creation

v1.32.2   2014-12-22 09:24:36-08:00 America/Los_Angeles

	+ Added better error messages when build_instance fails
	+ Closed #1 in github by converting from Smart::Comments to Log::Shiras debug reporting!

v1.28.6   2014-11-26 09:19:45-08:00 America/Los_Angeles

	! Explicitly added all pre-reqs to the test phase to pass a clean build

v1.28.4   2014-11-24 11:44:50-08:00 America/Los_Angeles

	! Fixed the version to number based for systems that hate 'v'

v1.28.2   2014-10-03 08:15:15-07:00 America/Los_Angeles

	! Fixed heading level on POD
	! Added minimum Moose dependancy to eliminate CPAN testers fails

v1.26.2   2014-07-28 17:51:16-07:00 America/Los_Angeles

	! Removed the no Type::Tiny::XS since it wasn't in a differnt package that 
		passed while this one failed

v1.24.2   2014-07-28 09:06:27-07:00 America/Los_Angeles

	! Yet another attempt to fix the non zero wait status 139 segfault errors
		(On a non XS module no less)

v1.22.6   2014-07-27 07:25:31-07:00 America/Los_Angeles

	! Figured out how to turn off Type::Tiny::XS altogether in testing so that 
		Failures are constrained to issues with this package only (I hope)

v1.22.4   2014-07-26 11:13:53-07:00 America/Los_Angeles

	! Fixed missing test plans

v1.22.2   2014-07-25 23:45:41-07:00 America/Los_Angeles

	! Required Type::Tiny Version 0.046 to settle on error messages
	! Required Type::Tiny::XS version 0.010 in testing but not installation

v1.20.2   2014-07-24 19:43:22-07:00 America/Los_Angeles

	! Fix fails associated with Type::Tiny::XS
		Really just caused the module to fail if Type::Tiny::XS 
		is loaded with a version earlier than 0.010

v1.18.2   2014-07-21 08:17:42-07:00 America/Los_Angeles

	! Reqired Type::Tiny 0.046 since 0.0.45_3 is broken
	! Made a few POD changes for clarification
	! Fixed the load test to clarify the requirements

v1.16.2   2014-07-10 16:27:41-07:00 America/Los_Angeles

	! another, another try at fixing the META files

v1.14.2   2014-07-10 16:05:44-07:00 America/Los_Angeles

	! another try at fixing the repository link in the META files

v1.12.2   2014-07-10 07:25:52-07:00 America/Los_Angeles

	! fixed the repository link in the META files

v1.10.2   2014-06-09 08:35:00-07:00 America/Los_Angeles

	! checked for mutability when building/adding roles and attributes
		and force mutability if not available

v1.8.2    2014-06-03 04:35:45-07:00 America/Los_Angeles

	+ updated the README to include a suggested tar unzip string
	+ switched the load order of add_roles_in_sequence and add_attributes and add_methods
		The goal is to let the roles also 'require' 'ad_hoc' stand alone methods and methods from attributes

v1.6.2    2014-05-30 18:46:44-07:00 America/Los_Angeles

	! fixed the build_class documentation
		changed the second 'add_roles_in_sequence' to 'add_methods'
	- changed $instance_count to $anonymous_class_count
	! changed our $anonymous_class_count to start at 0 since there isn't one at startup
	+ added a test for $anonymous_class_count

v1.4.2    2014-05-28 17:12:29-07:00 America/Los_Angeles

	! changed use version 0.94 to use version to fix CPAN testers fails
	! changed types test to accept a broader range of error messages for NameSpace fails

v1.2.2    2014-05-27 13:26:13-07:00 America/Los_Angeles


t/Mineral.pm
package Mineral;
use Moose;
use MooseX::StrictConstructor;

has 'type_of_mineral' =>( is => 'ro' );

no Moose;
__PACKAGE__->meta->make_immutable;

1;

t/Identity.pm
package Identity;
use Moose::Role;
use MooseX::StrictConstructor;
requires 'is_important';

has 'name' =>( 
		is => 'ro',
		#~ reader => 'get_name',
	);
	
no Moose::Role;
1;

t/Vegetable.pm
package Vegetable;
use Moose;
use MooseX::StrictConstructor;

has 'type_of_mineral' =>( is => 'ro' );

no Moose;
__PACKAGE__->meta->make_immutable;

1;

lib/MooseX/Util/ClassBuilder.pm
package MooseX::Util::ClassBuilder;
use 5.010;
use Moose;
use Moose::Meta::Class;
use version; our $VERSION = qv('0.001_001');
use Moose::Exporter;
Moose::Exporter->setup_import_methods(
	as_is => [ 'build_class', 'build_instance' ],
);

BEGIN{
	if( $ENV{ Smart_Comments } ){
		use Smart::Comments -ENV;
		### Smart-Comments turned on for MooseX-Util-ClassBuilder ...
	}
}

###############  Package Variables  #####################################################

our	$instance_count //= 1;

###############  Dispatch Tables  #######################################################

###############  Public Attributes  #####################################################

###############  Public Methods  ########################################################

sub build_class{
	### <where> - reached build_class ...
	##### <where> - passed arguments: @_
	my	$args = ( scalar( @_ ) == 1 ) ? $_[0] : { @_ };
	my ( $name, $class_args );
	if( exists $args->{class_name} ){
		$name = $args->{class_name};
		delete $args->{class_name};
	}else{
		$name = "ANONYMOUS_SHIRAS_MOOSE_CLASS_" . $instance_count++;
	}
	if( exists $args->{superclasses} ){
		$class_args->{superclasses} = $args->{superclasses};
		delete $args->{superclasses};
	}
	if( exists $args->{roles} ){
		$class_args->{roles} = $args->{roles};
		delete $args->{roles};
	}
	my $want_array = ( caller(0) )[5];
	### <where> - name: $name
	### <where> - class args: $class_args
	### <where> - remaining arguments: $args
	### <where> - want array: $want_array
	my 	$class_name =	Moose::Meta::Class->create(
							$name, %{$class_args},
						)->name;
	### <where> - returning the name: $class_name
	if( $want_array ){
		return ( $class_name, $args );
	}else{
		return $class_name;
	}
}

sub build_instance{
	my	$args = { @_ };
	### <where> - reached build_instance ...
	##### <where> - passed arguments: $args
	my ( $class_name, $instance_args ) = build_class( $args );
	my	$instance = $class_name->new( $instance_args );
	##### <where> - instance: $instance
	return $instance;
}

###############  Private Attributes  ####################################################

###############  Private Methods / Modifiers  ###########################################

#################### Phinish with a Phlourish ###########################################

no Moose;
__PACKAGE__->meta->make_immutable;

1;
# The preceding line will help the module return a true value

#################### main pod documentation begin #######################################

__END__

=head1 NAME

MooseX::Util::ClassBuilder - Yet another way to build Moose Classes

=head1 SYNOPSIS
    
	#!perl
	use Modern::Perl;

	package Mineral;
	use Moose;

	has 'type' =>( is => 'ro' );

	package Identity;
	use Moose::Role;

	has 'name' =>( is => 'ro' );

	use MooseX::Util::ClassBuilder qw( build_class );
	use Test::More;
	use Test::Moose;

	my 	$pet_rock_class = build_class(
			class_name => 'Pet::Rock',
			superclasses =>['Mineral'],
			roles =>['Identity'],
		);

	my 	$paco = $pet_rock_class->new(
			type => 'Quartz',
			name => 'Paco',
		);

	does_ok( $paco, 'Identity', 'Check that the ' . $paco->meta->name . ' has an -Identity-' );
	say 'My ' . $paco->meta->name . ' made from -' . $paco->type . '- (a ' .
	( join ', ', $paco->meta->superclasses ) . ') is called -' . $paco->name . '-';
	done_testing();
    
    ############################################################################
    #     Output of SYNOPSIS
    # 01:ok 1 - Check that the Pet::Rock has an -Identity-
    # 02:My Pet::Rock made from -Quartz- (a Mineral) is called -Paco-
    # 03:1..1
    ##############################################################################

    
=head1 DESCRIPTION

This module is used to compose Moose classes and instances on the fly.

=head1 Methods

=head2 Exported Methods

=head3 build_class( %args|\%args )

=over

=item B<Definition:> This method is used to compose a Moose Class.  It will 
take the passed arguments and strip out three potential key value pairs.  It 
then uses the L<Moose::Meta::Class> module to build a new composed class.

=item B<Accepts:> a hash or hashref of arguments.  The three key value pairs 
use are;

=over

=item B<class_name> - This is the name (a string) that the new instance of 
a this class is blessed under.  If this key is not provided the package 
will generate a generic name.

=item B<superclasses> - this is intentionally the same key from 
L<Moose::Meta::Class>.  It expects the same values.

=item B<roles> - this is intentionally the same key from L<Moose::Meta::Class>.  
It expects the same values.

=back

=item B<Returns:> This will check the caller and see if it wants an array or a 
scalar.  In array context it returns the new class name and a hash ref of the 
unused hash key - value pairs.  These are presumably the arguments for the 
instance.  If the requested return is a scalar it just returns the name of 
the newly created class.

=back

=head3 build_instance( %args|\%args )

=over

=item B<Definition:> This method is used to create a Moose Class instance.  I<It 
assumes that you do not have the class pre-built and will look for the needed 
information to compose a new class as well.>  Basically this passes the %args 
intact to L<build_class|/build_class( %args|\%args )> and then runs 
$returned_class_name->new( %remaining_args );

=item B<Accepts:> a hash or hashref of arguments.  They must include the 
necessary information to build a class.  I<(if you already have a class just 
call $class-E<gt>new(); instead of this method!)> This hashref can also 
contain any attribute settings for the instance as well.


=item B<Returns:> This will return a blessed instance of your new class with 
the passed attributes set.

=back

=head1 GLOBAL VARIABLES

=over

=item B<$ENV{Smart_Comments}>

The module uses L<Smart::Comments> if the '-ENV' option is set.  The 'use' is 
encapsulated in a BEGIN block triggered by the environmental variable to comfort 
non-believers.  Setting the variable $ENV{Smart_Comments} will load and turn 
on smart comment reporting.  There are three levels of 'Smartness' available 
in this module '### #### #####'.

=back

=head1 SUPPORT

=over

=item L<MooseX-Util-ClassBuilder/issues|https://github.com/jandrew/MooseX-Util-ClassBuilder/issues>

=back

=head1 TODO

=over

=item * Add a type package to manage the inputs to the exported methods

=back

=head1 AUTHOR

=over

=item Jed Lund

=item jandrew@cpan.org

=back

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=head1 Dependencies

=over

=item L<version>

=item L<5.010> - use of defined or

=item L<Moose>

=item L<Moose::Meta::Class>

=item L<Moose::Exporter>

=back

=head1 SEE ALSO

=over

=item L<Moose::Util> - with_traits

=item L<MooseX::ClassCompositor>

=item L<Smart::Comments> - is used if the -ENV option is set

=back

=cut

#################### main pod documentation end #########################################


Makefile.PL
# This Makefile.PL for MooseX-ShortCut-BuildInstance was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.38.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.010;
use ExtUtils::MakeMaker 7.04;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A shortcut to build Moose instances",
  "AUTHOR" => "Jed Lund <jandrew\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "7.04",
    "Moose" => "2.1213"
  },
  "DISTNAME" => "MooseX-ShortCut-BuildInstance",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010",
  "NAME" => "MooseX::ShortCut::BuildInstance",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Clone" => 0,
    "Data::Dumper" => 0,
    "Moose" => "2.1213",
    "Moose::Exporter" => 0,
    "Moose::Meta::Class" => 0,
    "Moose::Util" => 0,
    "Type::Library" => 0,
    "Type::Utils" => "1.000",
    "Types::Standard" => "1.000",
    "lib" => 0,
    "strict" => 0,
    "utf8" => 0,
    "version" => "0.77",
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => "0.12",
    "Carp" => 0,
    "Data::Dumper" => 0,
    "DateTime" => 0,
    "List::Util" => "1.33",
    "Moose" => "2.1213",
    "Moose::Exporter" => 0,
    "Moose::Meta::Class" => 0,
    "Moose::Role" => 0,
    "Moose::Util" => 0,
    "MooseX::HasDefaults::RO" => 0,
    "MooseX::StrictConstructor" => 0,
    "Smart::Comments" => 0,
    "Test::Moose" => 0,
    "Test::Most" => 0,
    "Type::Tiny" => "1.000",
    "lib" => 0,
    "version" => "0.77"
  },
  "VERSION" => "v1.42.0",
  "test" => {
    "TESTS" => "t/MooseX/ShortCut/*.t t/MooseX/ShortCut/BuildInstance/*.t"
  }
);

my %FallbackPrereqs = (
  "Capture::Tiny" => "0.12",
  "Carp" => 0,
  "Clone" => 0,
  "Data::Dumper" => 0,
  "DateTime" => 0,
  "List::Util" => "1.33",
  "Moose" => "2.1213",
  "Moose::Exporter" => 0,
  "Moose::Meta::Class" => 0,
  "Moose::Role" => 0,
  "Moose::Util" => 0,
  "MooseX::HasDefaults::RO" => 0,
  "MooseX::StrictConstructor" => 0,
  "Smart::Comments" => 0,
  "Test::Moose" => 0,
  "Test::Most" => 0,
  "Type::Library" => 0,
  "Type::Tiny" => "1.000",
  "Type::Utils" => "1.000",
  "Types::Standard" => "1.000",
  "lib" => 0,
  "strict" => 0,
  "utf8" => 0,
  "version" => "0.77",
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
